### PR TITLE
Add support for Pascal Context 59 classes

### PR DESCRIPTION
**Pull request by sennnnn**
* Create PascalContextDataset59 class in mmseg/datasets/pascal_context.py;

* Set reduce_zero_label=True for train_pipeline and PascalContextDataset59;
    
* Add some configs for Pascal-Context 59 classes training and testing;

All evaluations are conducted after PR #471.

This table aims at checking the correctness of pascal context 59 dataset support.

| model | test repo | num classes | Lr schd | mIoU | mIoU(ms+flip) |
| :-- | :--: | :--: | :--: | :--: | :--: |
| HRNetV2-W48(author) | mmsegmentation | 59 | - | 51.7 | 53.92 |
| HRNetV2-W48(author) | mmsegmentation | 60 | - | 46.18 | 48.02 |
| HRNetV2-W48(author) | raw | 59 | - | 52.25 | 54.1 |
| HRNetV2-W48(author) | raw | 60 | - | 46.96 | 48.3 |
| HRNetV2-W48(new) | mmsegmentation | 59 |40000  | 50.37 | 52.88 |
| HRNetV2-W48(new) | mmsegmentation | 60 | 40000 | 45.55 | 47.7 |
| HRNetV2-W48(new) | mmsegmentation | 59 | 80000 | 51.12 | 53.56 |
| HRNetV2-W48(new | mmsegmentation | 60 | 80000 | 45.79 | 47.86 |
| HRNetV2-W48(old) | mmsegmentation | 60 | 40000 | 45.14 | 47.42 |

This table is the training and evaluation collection.

| model  | num classes | Lr schd | mIoU | mIoU(ms+flip) |
| :-- | :--: | :--: | :--: | :--: |
| DeepLabV3+_R-101-D8 | 59 | 40000 | 52.86 | 54.54 |
| DeepLabV3+_R-101-D8 | 59 | 80000 | 53.2 | 54.67 |
| DeepLabV3_R-101-D8 | 59 | 40000 | 52.61 | 54.28 |
| DeepLabV3_R-101-D8 | 59 | 80000 | 52.46 | 54.09 |
| PSPNet_R-101-D8 | 59 | 40000 | 52.02 | 53.54 |
| PSPNet_R-101-D8 | 59 | 80000 | 52.47 | 53.99 |
| HRNetV2-W48 | 59 | 40000 | 50.33 | 52.83 |
| HRNetV2-W48 | 59 | 80000 | 51.12 | 53.56 | 
| FCN_R-101-D8 | 59 | 40000 | 48.42 | 50.4 |
| FCN_R-101-D8 | 59 | 80000 | 49.35 | 51.38 |

# How to use [HRNetV2-W48 official weights](https://github.com/HRNet/HRNet-Semantic-Segmentation/tree/master) to test on [our repo](https://github.com/open-mmlab/mmsegmentation)?

```
ps: testing on pascal context dataset
```

1. Download HRNetV2-W48 official weights [59 classes](https://1drv.ms/u/s!Aus8VCZ_C_33f5Bfbt4KmLeX8uw) and [69 classes](https://1drv.ms/u/s!Aus8VCZ_C_33gQEHDQrZCiv4R5mf). (You'd better download them in [master branch of official repo](https://github.com/HRNet/HRNet-Semantic-Segmentation/tree/master))
2. Then, using the convert script to convert official weights to pytorch style:

```
python tools/scripts/convert_hrnet.py [src_path] [dst_path]
```

3. In order to match the inference process, modify codes according to [this commit](https://github.com/sennnnn/mmsegmentation/commit/803dca1007ae745129317b4af2602a28f888355c).
4. Then, using hrnet config to test model on pascal context dataset:

```
python -u tools/test.py [config path] [converted official weights path] --eval mIoU
```

### CODE REVIEW

File: mmseg/datasets/pascal_context.py [hunk 1]
```diff
@@ -52,3 +52,52 @@ def __init__(self, split, **kwargs):
...
+               [250, 10, 15], [20, 255, 0], [31, 255, 0], [255, 31, 0],
+               [255, 224, 0], [153, 255, 0], [0, 0, 255], [255, 71, 0],
+               [0, 235, 255], [0, 173, 255], [31, 0, 255]]
+
+    def __init__(self, split, **kwargs):
+        super(PascalContextDataset59, self).__init__(
+            img_suffix='.jpg',
+            seg_map_suffix='.png',
+            split=split,
+            reduce_zero_label=False,
```
Comment (xvjiarui) [inline mmseg/datasets/pascal_context.py:101]: Should it be true?

File: configs/_base_/datasets/pascal_context_59.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+data_root = 'data/VOCdevkit/VOC2010/'
+img_norm_cfg = dict(
+    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
+
+img_scale = (520, 520)
+crop_size = (480, 480)
+
+train_pipeline = [
+    dict(type='LoadImageFromFile'),
+    dict(type='LoadAnnotations'),
```
Comment (xvjiarui) [inline configs/_base_/datasets/pascal_context_59.py:13]: May also need `reduce_zero_label`.

File: configs/_base_/datasets/pascal_context_59.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+data = dict(
+    samples_per_gpu=4,
+    workers_per_gpu=4,
+    train=dict(
+        type=dataset_type,
+        data_root=data_root,
+        img_dir='JPEGImages',
+        ann_dir='SegmentationClassContext',
+        split='ImageSets/SegmentationContext/train.txt',
+        pipeline=train_pipeline),
```
Comment (xvjiarui) [inline configs/_base_/datasets/pascal_context_59.py:47]: You need to pass the `reduce_zero_label` here.

File: mmseg/core/evaluation/metrics.py [hunk 4]
```diff
@@ -57,11 +57,11 @@ def intersect_and_union(pred_label,
 
     intersect = pred_label[pred_label == label]
     area_intersect = torch.histc(
-        intersect.float(), bins=(num_classes), min=0, max=num_classes)
+        intersect.float(), bins=num_classes, min=0, max=num_classes - 1)
```
Comment (xvjiarui) [inline mmseg/core/evaluation/metrics.py:60]: We may need to test whether the results are consistent for other models. 
We could make another PR for this modification, we may also include impact range in the new PR.

File: .gitignore [hunk 5]
```diff
@@ -108,10 +108,16 @@ data
...
+*.npy
 *.pkl
 *.pkl.json
 *.log.json
 work_dirs/
+backup/
+
+# temp file or folder
+/temp*/
+/temp*.*
```
Comment (xvjiarui) [inline .gitignore:120]: These could be removed in the final version.

File: configs/_base_/schedules/schedule_20k_epochwise.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# optimizer
+optimizer = dict(type='SGD', lr=0.01, momentum=0.9, weight_decay=0.0005)
+optimizer_config = dict()
+# learning policy
+lr_config = dict(policy='poly', power=0.9, min_lr=1e-4, by_epoch=True)
+# runtime settings
+runner = dict(type='IterBasedRunner', max_iters=20000)
+checkpoint_config = dict(by_epoch=True, interval=2000)
+evaluation = dict(interval=2000, metric='mIoU')
```
Comment (xvjiarui) [inline configs/_base_/schedules/schedule_20k_epochwise.py:9]: This config may not be necessary.

File: mmseg/core/evaluation/metrics.py (hunk 4 above)
Comment (xvjiarui) [reply mmseg/core/evaluation/metrics.py:60]: We may merge master and resolve conflicts.

File: configs/_base_/schedules/schedule_20k_epochwise.py (hunk 6 above)
Comment (sennnnn) [reply configs/_base_/schedules/schedule_20k_epochwise.py:9]: ok, I will clean these codes in next commit.

**Comment by xvjiarui** [comment]
Please merge master to resolve the conflicts.

**Comment by xvjiarui** [comment]
We may delete unused configs.

File: tools/test.py [hunk 7]
```diff
@@ -85,6 +90,9 @@ def main():
     cfg = mmcv.Config.fromfile(args.config)
     if args.options is not None:
         cfg.merge_from_dict(args.options)
+    # set transmit port
+    if args.port != '25900':
+        cfg.dist_params['port'] = args.port
```
Comment (xvjiarui) [inline tools/test.py:95]: These may not be necessary.

File: configs/hrnet/fcn_hr18s_480x480_40k_pascal_context_59.py [hunk 8]
```diff
@@ -0,0 +1,9 @@
+_base_ = './fcn_hr18_480x480_40k_pascal_context_59.py'
+model = dict(
+    pretrained='open-mmlab://msra/hrnetv2_w18_small',
+    backbone=dict(
+        extra=dict(
+            stage1=dict(num_blocks=(2, )),
+            stage2=dict(num_blocks=(2, 2)),
+            stage3=dict(num_modules=3, num_blocks=(2, 2, 2)),
+            stage4=dict(num_modules=2, num_blocks=(2, 2, 2, 2)))))
```
Comment (xvjiarui) [inline configs/hrnet/fcn_hr18s_480x480_40k_pascal_context_59.py:9]: These configs should be deleted.

**Review by xvjiarui — APPROVED** [review]
